Add CORB, CORP and Partitioned Caches Defense Articles to the Wiki (#6)
xsleaks/wiki: content/docs/defenses/browser-intrinsic/partitioned-cache.md

+++
title = "Partitioned HTTP Cache"
description = ""
date = "2020-07-21"
category = "defenses"
menu = "main"
+++

## Explanation

[Cache probing attacks](https://TODO) have been present on the web for a long time mainly because browsers HTTP cache is shared across all the websites visited by a user, allowing attackers to interact with it and infer private information from other origins.

<!--TODO(manuelvsousa): Add reference to cache probing attacks-->

> Considering Opt-in solutions, applications can use the [`Vary` Header combined with `Fetch-Metadata`](https://TODO) to restrict the cache usage to a certain group of origins or use [some workarounds]({{< ref "../design-protections/subresource-protections.md" >}}) to protect resources, which come with tradeoffs. Browsers, however, have been planning a defense mechanism to segregate the cached resources per origin/site, making it impossible for attackers pages to interact with cached contents of different origins[^1] [^2] [^3]. Specifically, browsers tested caching with double keys such as `top-frame origin` and URL, however, other keys and strategies (triple keys) have also been considered.

Reviewer: ```suggestion
Considering Opt-in solutions, applications can use the [`Vary` Header combined with `Fetch-Metadata`](https://TODO) to prevent cross-origin fetches to be affected by the site cache or use [some workarounds]({{< ref "../design-protections/subresource-protections.md" >}}) to protect resources, which come with downsides. Browsers, however, have been planning a defense mechanism to segregate the cached resources per origin/site, making it impossible for attackers pages to interact with cached contents of different origins[^1] [^2] [^3]. Specifically, browsers tested caching with finer grained keys such as the tuple <`top-frame origin`, URL> or even triples of values.
```

Author: I also replaced finer grained with finer-grained (3b4c9af)

> [Chrome](https://bugs.chromium.org/p/chromium/issues/detail?id=910708) and [Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1590107) are still discussing an implementation for a Multi-Keyed Cache while [Safari](https://bugs.webkit.org/show_bug.cgi?id=110269) already runs a version of it since 2013.

Reviewer: I would change the links to be smth like 

Chrome [bug](url)

Otherwise it looks like the "Chrome" link is a link to the browser itself.

Author: Saying "bug" only to have a link seems odd. Maybe we could reference: Chrome (bug) ?.

I followed a similar approach applied in the [clocks article](https://xsleaks.github.io/docs/attacks/timing-attacks/clocks/#performancenow-api). What do you think of this? (Applied in e59ac50)

> <!--TODO(manuelvsousa): Add socket exhaustion ?-->

Reviewer: Well, not as part of the partitioned cache.

Author: I don't know why this is here actually :/

Author: Fixed in 82d8aa6

## Relevant Projects

### Intelligent Tracking Prevention

[Intelligent Tracking Prevention](https://webkit.org/tracking-prevention/) (ITP) is a privacy feature part of WebKit. It's a conjunction of several tracking prevention features containing a partitioned HTTP Cache which originated from the initial [WebKit Keyed cache from 2013](https://bugs.webkit.org/show_bug.cgi?id=110269). The used keys for the cache are the top frame's eTLD+1 and the origin of each fetched subresource.

{{< hint warning >}}

> Researchers found out attackers [could abuse some ITP features](https://TODO) to infer to private user data.

Reviewer: ```suggestion
Researchers found out attackers [could abuse some ITP features](https://TODO) to infer private user data.
```

{{< /hint >}}

<!--TODO(manuelvsousa): ADD ITP XS-Leak here -->

### First Party Isolation

First Party Isolation is a [Browser Extension](https://addons.mozilla.org/en-US/firefox/addon/first-party-isolation/) for Firefox which restricts access to cookies and persistent data (e.g cache) per domain.

## Considerations

Partitioned HTTP caches are a promising security feature that will eventually land in browsers. These partitioning strategies will mitigate all the XS-Leaks leveraging browsers caches and might be extended to other browser resources which help mitigate resilient attack vectors like the [Socket Exhaustion XS-Leak](https://TODO).

> <!--TODO(manuelvsousa): Add socket exhaustion ?-->

Reviewer: Here it might be ok to add a reference, but at that point I would change the title to be "Partitioned browser resources".

Author: I don't think we want to change the article to Partitioned browser resources unless we gather more knowledge about it. The only reference we have about partitioning the network pool is a github comment, so I don't really know how committed browsers are into shipping such features. You know anything from our side related to this?

Well, I mentioned the partitioning proposal of network pools in the Socket Exhaustion XS-Leak , so maybe we should remove this from this article?

Reviewer 2: I found a tracking [bug#1572544](https://bugzilla.mozilla.org/show_bug.cgi?id=1572544) in Firefox for *Partition our HTTP connection pool per the top-level document's origin* so I would say this is in motion.

Author: Yup, I think we refer this in the socket exhaustion article. If not I will add it

+++
title = "Partitioned HTTP Cache"
description = ""
date = "2020-07-21"
category = "defenses"
menu = "main"
+++

## Explanation

> [Cache probing attacks](https://TODO) have been present on the web for a long time mainly because browsers HTTP cache is shared across all the websites visited by a user, allowing attackers to interact with it and extract private information from other origins.

Reviewer: ```suggestion
[Cache probing attacks](https://TODO) have been present on the web for a long time mainly because browsers HTTP cache is shared across all the websites visited by a user, allowing attackers to interact with it and infer private information from other origins.
```

<!--TODO(manuelvsousa): Add reference to cache probing attacks-->

> There is not much applications can do stop these attacks, [apart for some workarounds]({{< ref "../design-protections/subresource-protections.md" >}}) which come with tradeoffs. Browsers, however, have been planning a defense mechanism to segregate the cached resources per origin/site, making it impossible to attackers pages to interact with cached contents of different origins[^1] [^2] [^3]. Specifically, browsers tested caching with double keys such as `top-frame origin` and URL, however, other keys and strategies (triple keys) have also been considered.

Reviewer: What about setting `Vary: Sec-Fetch-Site` ?

Author: Yes! This is very good. I will add it

Author: I referenced this in the article (a7de532) but I will make a proper section about it in the cache probing attack article and maybe fetch-metadata.

> There is not much applications can do stop these attacks, [apart for some workarounds]({{< ref "../design-protections/subresource-protections.md" >}}) which come with tradeoffs. Browsers, however, have been planning a defense mechanism to segregate the cached resources per origin/site, making it impossible to attackers pages to interact with cached contents of different origins[^1] [^2] [^3]. Specifically, browsers tested caching with double keys such as `top-frame origin` and URL, however, other keys and strategies (triple keys) have also been considered.

Reviewer: ```suggestion
There is not much applications can do stop these attacks, [apart for some workarounds]({{< ref "../design-protections/subresource-protections.md" >}}) which come with tradeoffs. Browsers, however, have been planning a defense mechanism to segregate the cached resources per origin/site, making it impossible for attackers pages to interact with cached contents of different origins[^1] [^2] [^3]. Specifically, browsers tested caching with double keys such as `top-frame origin` and URL, however, other keys and strategies (triple keys) have also been considered.
```

{{< hint info >}}

> Chrome [^1] and Firefox [^2] are still discussing an implementation for a Multi-Keyed Cache while Safari [^3] already runs a version of it since 2013.

Reviewer 2: Cache probing does not seem to be working in Firefox anymore. I think this is because of the fix to [bug#1536058](https://bugzilla.mozilla.org/show_bug.cgi?id=1536058)

Author: I was aware of this bug and I just noticed it's marked as fixed! I thought that apart from Safari, nobody had pushed to stable this feature. I had to confirm this and I now have to say, **I could not reproduce the cache partition**.

There is a flag in `about:config` named `browser.cache.cache_isolation` which is `false` by default. Changing to true and testing again, the partitioned cache works, and shows me the behaviour I am expecting! Really weird. Maybe the issue just says they pushed it to stable but under a perf flag?

What do you think. Can you take a look at this?

Reviewer 2: I think you might be right that it's under perf flag. From the [code](https://hg.mozilla.org/mozilla-central/rev/a5e791146ef5#l2.15)

```java
+  mIsIsolated = StaticPrefs::browser_cache_cache_isolation() ||
+                (IsThirdPartyTrackingResource() &&
+                 !AntiTrackingCommon::IsFirstPartyStorageAccessGrantedFor(
+                     this, mURI, nullptr));
```

I can dig up on it more, last time I tried XS-Search on Firefox, I couldn't make it working. I was evicting resource successfully for the 'malicious' origin, but when loaded in the iframe, it used different cache, so the evicting did nothing.

Reviewer 2: Leave as it is now until we confirm what Firefox does.

> [Chrome](https://bugs.chromium.org/p/chromium/issues/detail?id=910708) and [Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1590107) are still studying an implementation for a Keyed Cache while [Safari](https://bugs.webkit.org/show_bug.cgi?id=110269) already runs a version of it since 2013.

Reviewer: ```suggestion
[Chrome](https://bugs.chromium.org/p/chromium/issues/detail?id=910708) and [Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1590107) are still discussing an implementation for a Multi-Keyed Cache while [Safari](https://bugs.webkit.org/show_bug.cgi?id=110269) already runs a version of it since 2013.
```

<!--TODO(manuelvsousa): Add socket exhaustion ?-->

## Relevant Projects

### Intelligent Tracking Prevention

> [Intelligent Tracking Prevention](https://webkit.org/tracking-prevention/) (ITP) is a privacy feature part of WebKit. It's a conjunction of several tracking prevention features against malicious trackers, containing a partitioned HTTP Cache which originated from the initial [WebKit Keyed cache from 2013](https://bugs.webkit.org/show_bug.cgi?id=110269). The used keys for the cache are the top frame's eTLD+1 and the origin of each fetched subresource.

Reviewer: ```suggestion
[Intelligent Tracking Prevention](https://webkit.org/tracking-prevention/) (ITP) is a privacy feature part of WebKit. It's a conjunction of several tracking prevention features containing a partitioned HTTP Cache which originated from the initial [WebKit Keyed cache from 2013](https://bugs.webkit.org/show_bug.cgi?id=110269). The used keys for the cache are the top frame's eTLD+1 and the origin of each fetched subresource.
```

{{< hint warning >}}

> Several XS-Leaks resulted in browser features triggering changes in fundamental web platform features. One of those XS-Leaks abused ITP features which could lead to private information leaks.

Reviewer: This sentence is very hard to read, can you please try to use less passive and subordinates and be more assertive?

Author: Fixed in b740814

Reviewer: "infer to" in that commit is not correct, I would say "exfiltrate" or just "infer" (with no "to).

Author: Sorry it was a typo. This was fixed with your suggestion f3d2b3b8e7dbe7d10bf24e3560ca189a641cbc85

{{< /hint >}}

<!--TODO(manuelvsousa): ADD ITP XS-Leak here -->

### First Party Isolation

First Party Isolation is a [Browser Extension](https://addons.mozilla.org/en-US/firefox/addon/first-party-isolation/) for Firefox which restricts access to cookies and persistent data (e.g cache) per domain.

## Considerations

> Partitioned HTTP caches are a promising security feature that will eventually land in browsers. These partitioning strategies will mitigate all the XS-Leaks leveraging browsers caches and might spread to other browser resources which help mitigate resilient attack vectors like the [Socket Exhaustion XS-Leak](https://TODO).

Reviewer: ```suggestion
Partitioned HTTP caches are a promising security feature that will eventually land in browsers. These partitioning strategies will mitigate all the XS-Leaks leveraging browsers caches and might be extended to other browser resources which help mitigate resilient attack vectors like the [Socket Exhaustion XS-Leak](https://TODO).
```

First Party Isolation is a [Browser Extension](https://addons.mozilla.org/en-US/firefox/addon/first-party-isolation/) for Firefox which restricts access to cookies and persistent data (e.g cache) per domain.

## Considerations

Partitioned HTTP caches are a promising security feature that will eventually land in browsers. These partitioning strategies will mitigate all the XS-Leaks leveraging browsers caches and might be extended to other browser resources which help mitigate resilient attack vectors like the [Socket Exhaustion XS-Leak](https://TODO).

<!--TODO(manuelvsousa): Add socket exhaustion ?-->

> ## XS-Leaks Mitigation Overview

Reviewer 2: I don't understand the table below. Why did we only mention three XS-Leaks here? Is a `<!--TODO>` missing?

An introduction to this table would be also helpful. It's not obvious what it is trying to show.

Author: Yup, I think we refer this in the socket exhaustion article. If not I will add it

Author: Yes this is incomplete. We have to swipe the defense articles and add attacks we present in the wiki. I will add this to the long term todos.

Reviewer 2: Can you add a `<!--TODO` under a table to clarify that it needs further improvements? It's OK I guess to leave it as it is now, as long as we know it's not finished yet. I would still like to see a one or two sentences introducing to the table.

Author: We will leave as it is. All defense mechanisms are incomplete by now

Reviewer 2: Can you add this to long term TODOs then?

| [ID Leaks]({{< ref "../../attacks/id-attribute.md" >}})           |         ❌                   |         ❌

> ### Table Caption

Reviewer: ```suggestion
### Table Legend
```

| [Frame Counting]({{< ref "../../attacks/frame-counting.md" >}})   |         ❌                   |         ❌
| [Navigations]({{< ref "../../attacks/navigations.md" >}})         |         ❌                   |         ❌
| [ID Leaks]({{< ref "../../attacks/id-attribute.md" >}})           |         ❌                   |         ❌

### Table Legend

> - **Partitioned Cache Protection** - At least one attack alternative of the XS-Leak can be mitigated if the defense mechanism would exist

Reviewer: ```suggestion
- **Partitioned Cache Protection** - At least one attack alternative of the XS-Leak can be mitigated
```

> - **Full Mitigation** - Fully mitigates an XS-Leak if the defense mechanism would exist

Reviewer: ```suggestion
- **Full Mitigation** - Fully mitigates an XS-Leak
```
